Limits of sequences in metric spaces (#1378)
UniMath/agda-unimath: src/metric-spaces/limits-sequences-pseudometric-spaces.lagda.md

open import elementary-number-theory.natural-numbers
open import elementary-number-theory.positive-rational-numbers

open import foundation.asymptotically-constant-sequences
open import foundation.asymptotically-equal-sequences
open import foundation.constant-maps
open import foundation.dependent-pair-types
open import foundation.functoriality-dependent-pair-types
open import foundation.propositions
open import foundation.sequences
open import foundation.subsequences
open import foundation.transport-along-identifications
open import foundation.universe-levels

open import metric-spaces.limits-sequences-premetric-spaces
open import metric-spaces.pseudometric-spaces
open import metric-spaces.sequences-pseudometric-spaces
```

</details>

## Ideas

{{#concept "Limits" Disambiguation="of sequences in pseudometric spaces" Agda=is-limit-sequence-Pseudometric-Space}}
of
[sequences in pseudometric spaces](metric-spaces.sequences-pseudometric-spaces.md)
are [limits](metric-spaces.limits-sequences-premetric-spaces.md) in the
underlying [premetric space](metric-spaces.premetric-spaces.md).

Limits of a sequence in a pseudometric space are indistinguishable. The

> (asymptotical) value of an

Reviewer: Maybe "The asymptote of an asymptotically constant sequence..."

([asymptotically](foundation.asymptotically-constant-sequences.md)) constant
sequence in a pseudometric is a limit of the sequence.

> Asymptotical indistinguishability of sequences in pseudometric spaces preserves

Reviewer: Definitely just "Asymptotic" here

    is-limit-sequence-Premetric-Space
      ( premetric-Pseudometric-Space M)
      ( u)
      ( l)

  modulus-limit-sequence-Pseudometric-Space :
    is-limit-sequence-Pseudometric-Space → ℚ⁺ → ℕ
  modulus-limit-sequence-Pseudometric-Space =
    modulus-limit-sequence-Premetric-Space
      ( premetric-Pseudometric-Space M)
      ( u)
      ( l)

  is-modulus-modulus-limit-sequence-Pseudometric-Space :
    (H : is-limit-sequence-Pseudometric-Space) (d : ℚ⁺) →
    (n : ℕ) (I : leq-ℕ (modulus-limit-sequence-Pseudometric-Space H d) n) →
    neighborhood-Pseudometric-Space M d (u n) l
  is-modulus-modulus-limit-sequence-Pseudometric-Space =
    is-modulus-modulus-limit-sequence-Premetric-Space
      ( premetric-Pseudometric-Space M)
      ( u)
      ( l)
```

## Properties

### Constant sequences in pseudometric spaces have limit equal to their value

```agda
module _
  {l1 l2 : Level} (M : Pseudometric-Space l1 l2) (x : type-Pseudometric-Space M)
  where

  limit-constant-sequence-Pseudometric-Space :
    is-limit-sequence-Pseudometric-Space M (const ℕ x) x
  limit-constant-sequence-Pseudometric-Space d =
    zero-ℕ , λ _ _ → is-reflexive-structure-Pseudometric-Space M d x
```

> ### Asymptotical values of asymptotically constant sequences in pseudometric spaces are limits

Reviewer: "Asymptotic values," or "the asymptote"

Author: I went with asymptotic values. If you believe asymptote is better, I can rename `value-asymptotically-constant-sequence` into `asymptote-asymptotically-constant-sequence` and use this terminology everywhere.

Author: Now "Eventual value of eventually constant sequences ..."

open import elementary-number-theory.maximum-natural-numbers
open import elementary-number-theory.natural-numbers
open import elementary-number-theory.positive-rational-numbers

open import foundation.constant-maps
open import foundation.dependent-pair-types
open import foundation.functoriality-dependent-pair-types
open import foundation.inhabited-subtypes
open import foundation.inhabited-types
open import foundation.propositional-truncations
open import foundation.propositions
open import foundation.sequences
open import foundation.subsequences
open import foundation.subtypes
open import foundation.transport-along-identifications
open import foundation.universe-levels

open import metric-spaces.limits-sequences-premetric-spaces
open import metric-spaces.pseudometric-spaces
open import metric-spaces.sequences-pseudometric-spaces
```

</details>

## Ideas

{{#concept "Limits" Disambiguation="of sequences in pseudometric spaces" Agda=is-limit-sequence-Pseudometric-Space}}
of
[sequences in pseudometric spaces](metric-spaces.sequences-pseudometric-spaces.md)
are [limits](metric-spaces.limits-sequences-premetric-spaces.md) in the

> underlying [premetric space](metric-spaces.premetric-spaces.md).

Reviewer 2: please repeat the definition here

> Limits of a sequence in a pseudometric space are indistinguishable. The value of

Reviewer 2: Please define "indistinguishable", e.g., is this similarity with respect to the structure of a pseudometric/premetric? And give a link to the file where the concept is defined.

Author: *Indistinguishability* is a primary notion of [premetric structures](https://unimath.github.io/agda-unimath/metric-spaces.premetric-structures.html#3488), [premetric spaces](https://unimath.github.io/agda-unimath/metric-spaces.premetric-spaces.html#2395), [pseudometric spaces](https://unimath.github.io/agda-unimath/metric-spaces.pseudometric-spaces.html#4189) (where it's an equivalence relation) and [metric spaces](https://unimath.github.io/agda-unimath/metric-spaces.metric-spaces.html#7321) (where it's equivalent to equality). So it's defined a bit everywhere in the `metric-spaces` module.

Reviewer 2: Right, I suppose it might not be necessary to repeat the definition here, but it seems like it would be reasonable to have separate files about these concepts that we can then link to.

Author: > Right, I suppose it might not be necessary to repeat the definition here, but it seems like it would be reasonable to have separate files about these concepts that we can then link to.

Sure. If you'd like to make a rough sketch of what wish these modules to contain, I could make it work.

Reviewer 2: The following should be enough:

- For premetric spaces just give the definition. The idea section could be valuable for some discussion on the role of this concept, but that's not a requirement.
- For pseudometric spaces give the definition and show it is an equivalence relation. We might also add an "indistinguishability syntax".

Author: Ok. I think it's better to do it in another PR though. Feel free to make an Issue an assign me to it so I don't forget.

      ( u)
      ( l)

  is-modulus-limit-sequence-Pseudometric-Space : (ℚ⁺ → ℕ) → UU l2
  is-modulus-limit-sequence-Pseudometric-Space m =
    type-Prop (is-modulus-limit-prop-sequence-Pseudometric-Space m)

  modulus-limit-sequence-Pseudometric-Space : UU l2
  modulus-limit-sequence-Pseudometric-Space =
    type-subtype is-modulus-limit-prop-sequence-Pseudometric-Space

  modulus-modulus-limit-sequence-Pseudometric-Space :
    modulus-limit-sequence-Pseudometric-Space → ℚ⁺ → ℕ
  modulus-modulus-limit-sequence-Pseudometric-Space m = pr1 m

  is-modulus-modulus-limit-sequence-Pseudometric-Space :
    (m : modulus-limit-sequence-Pseudometric-Space) →
    is-modulus-limit-sequence-Pseudometric-Space
      (modulus-modulus-limit-sequence-Pseudometric-Space m)
  is-modulus-modulus-limit-sequence-Pseudometric-Space m = pr2 m

  is-limit-prop-sequence-Pseudometric-Space : Prop l2
  is-limit-prop-sequence-Pseudometric-Space =
    is-inhabited-subtype-Prop is-modulus-limit-prop-sequence-Pseudometric-Space

  is-limit-sequence-Pseudometric-Space : UU l2
  is-limit-sequence-Pseudometric-Space =
    type-Prop is-limit-prop-sequence-Pseudometric-Space
```

## Properties

> ### The value of a constant sequences in pseudometric spaces is a limit

Reviewer 2: ```suggestion
### The value of a constant sequence is its limit
```

Author: Note that this is not true in premetric spaces (you need reflexivity) and, in a pseudometric space, the limit may not be unique, so it feels weird to talk about *its* limit. Your suggestion loses a lot of these nuances. Are you absolutely sure about it?

Reviewer 2: > Note that this is not true in premetric spaces (you need reflexivity)

This is one of the reasons why I wouldn't call a premetric space a "premetric space", but rather a type equipped with (rational) neighborhoods.

But fine, I get your point to emphasize the indefinite article "a". However, the current formulation doesn't specify it is a limit of the constant sequence, and also there's a typo in "a constant sequences".

Author: > This is one of the reasons why I wouldn't call a premetric space a "premetric space", but rather a type equipped with (rational) neighborhoods.

Ok. I remember there was some debates about naming objects in the `metric-spaces` module. Maybe we should re-open the issue sometime? And since @lowasser also worked quite a bit with this module, maybe they'd like to join the conversation?

> But fine, I get your point to emphasize the indefinite article "a". However, the current formulation doesn't specify it is a limit of the constant sequence, and also there's a typo in "a constant sequences".

Fair point, I'll find a better formulation.

```agda
module _
  {l1 l2 : Level} (M : Pseudometric-Space l1 l2) (x : type-Pseudometric-Space M)
  where

  modulus-limit-constant-sequence-Pseudometric-Space :
    modulus-limit-sequence-Pseudometric-Space M (const ℕ x) x
  modulus-limit-constant-sequence-Pseudometric-Space =
    ( λ _ → zero-ℕ) ,
    ( λ ε _ _ → is-reflexive-structure-Pseudometric-Space M ε x)

  limit-constant-sequence-Pseudometric-Space :
    is-limit-sequence-Pseudometric-Space M (const ℕ x) x
  limit-constant-sequence-Pseudometric-Space =
    unit-trunc-Prop modulus-limit-constant-sequence-Pseudometric-Space
```

> ### Two limits of a sequence in a pseudometric space are indistinguishable

Reviewer 2: ```suggestion
### The limit of a sequence in a pseudometric space is essentially unique
```

Author: I don't remember we ever used the term "essentially unique" in the `metric-spaces` module. It feels weird to introduce it here.

Reviewer 2: Alright, yeah, you're probably right.

```agda
module _
  {l1 l2 : Level} (M : Pseudometric-Space l1 l2)
  (u : sequence-type-Pseudometric-Space M)
  (x y : type-Pseudometric-Space M)
  where

  indistinguishable-limit-modulus-limit-sequence-Pseudometric-Space :
    modulus-limit-sequence-Pseudometric-Space M u x →
    modulus-limit-sequence-Pseudometric-Space M u y →

>     is-indistinguishable-Pseudometric-Space M x y

Reviewer 2: If this is the concept of similarity in premetric spaces, which I assume it is, we usually use the name `sim-Pseudometric-Space M x y` for it

Author: see https://github.com/UniMath/agda-unimath/pull/1378#discussion_r2067437409, but I don't remember we ever called it `sim-Pseudometric-Space`. Maybe it would be better to introduce this name in a separate PR?

Reviewer 2: Indeed, let's discuss later. See for instance 

- https://unimath.github.io/agda-unimath/real-numbers.similarity-real-numbers.html
- https://unimath.github.io/agda-unimath/order-theory.similarity-of-elements-large-preorders.html

Although there doesn't seem to be that many other instances of this term being used as of yet. I posted an issue yesterday that is also tangentially related: #1418

            leq-ℕ N n →
            neighborhood-Pseudometric-Space M d₂ (u n) y) →
        neighborhood-Pseudometric-Space M (d₁ +ℚ⁺ d₂) x y
      tr-modulus-indistinguishable d₁ d₂ (Nx , Mx) (Ny , My) =
        ( is-triangular-structure-Pseudometric-Space M
          ( x)
          ( u (max-ℕ Nx Ny))
          ( y)
          ( d₁)
          ( d₂)
          ( My (max-ℕ Nx Ny) (right-leq-max-ℕ Nx Ny))
          ( is-symmetric-structure-Pseudometric-Space M d₁
            ( u (max-ℕ Nx Ny))
            ( x)
            ( Mx (max-ℕ Nx Ny) (left-leq-max-ℕ Nx Ny))))

  indistinguishable-limit-sequence-Pseudometric-Space :
    is-limit-sequence-Pseudometric-Space M u x →
    is-limit-sequence-Pseudometric-Space M u y →
    is-indistinguishable-Pseudometric-Space M x y
  indistinguishable-limit-sequence-Pseudometric-Space Lx Ly =
    rec-trunc-Prop
      ( is-indistinguishable-prop-Pseudometric-Space M x y)
      ( λ My →
        rec-trunc-Prop
          ( Π-Prop ℚ⁺ (λ d → structure-Pseudometric-Space M d x y))
          ( λ Mx →
            indistinguishable-limit-modulus-limit-sequence-Pseudometric-Space
              Mx
              My)
          ( Lx))
      ( Ly)
```

> ### Subsequences preserve limits in pseudometric spaces

Reviewer 2: ```suggestion
### Taking subsequences preserve limits in pseudometric spaces
```
